File: sdks/java/basys.sdk/backends/org/eclipse/basyx/aas/backend/connected/property/ConnectedCollectionProperty.java
package org.eclipse.basyx.aas.backend.connected.property;

import java.util.Collection;

import org.eclipse.basyx.aas.api.exception.ServerException;
import org.eclipse.basyx.aas.api.exception.TypeMismatchException;
import org.eclipse.basyx.aas.api.resources.ICollectionProperty;
import org.eclipse.basyx.aas.api.resources.PropertyType;
import org.eclipse.basyx.vab.core.proxy.VABElementProxy;

/**
 * Connects to a PropertySingleValued as specified by VWiD containing a
 * collection
 * 
 * @author schnicke
 *
 */
public class ConnectedCollectionProperty extends ConnectedProperty implements ICollectionProperty {

	public ConnectedCollectionProperty(String path, VABElementProxy proxy) {
		super(PropertyType.Collection, path, proxy);
	}

	@Override
	public void set(Collection<Object> collection) throws ServerException {
		try {
			getProxy().updateElementValue(constructPath("value"), collection);
		} catch (Exception e) {
			throw new ServerException(e.getClass().toString(), e.getMessage());
		}
	}

	@Override
	public void add(Object newValue) throws ServerException, TypeMismatchException {
		try {
			getProxy().createElement(constructPath("value"), newValue);
		} catch (Exception e) {
			throw new ServerException(e.getClass().toString(), e.getMessage());
		}
	}

	@Override
	public void remove(Object objectRef) throws ServerException {
		try {
			getProxy().deleteElement(constructPath("value"), objectRef);
		} catch (Exception e) {
			throw new ServerException(e.getClass().toString(), e.getMessage());
		}
	}

	@Override
	public Collection<Object> getElements() throws ServerException {
		return getCollection();
	}

	@Override
	public int getElementCount() throws ServerException {
		return getCollection().size();
	}

	@SuppressWarnings("unchecked")
	private Collection<Object> getCollection() {
		return (Collection<Object>) getProxy().readElementValue(constructPath("value"));
	}

}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/connected/submodelelement/property/ConnectedCollectionProperty.cpp
/*
 * ConnectedCollectionProperty.cpp
 *
 *      Author: wendel
 */

#include "ConnectedCollectionProperty.h"
#include "submodel/api/submodelelement/property/IProperty.h"

namespace basyx {
namespace submodel {

using namespace submodelelement::property;

ConnectedCollectionProperty::ConnectedCollectionProperty(std::shared_ptr<vab::core::proxy::IVABElementProxy> proxy) :
  ConnectedProperty(PropertyType::Collection, proxy)
{}

void ConnectedCollectionProperty::set(const basyx::object::object_list_t & collection) const
{
  this->getProxy()->updateElementValue(IProperty::Path::Value, collection);
}

void ConnectedCollectionProperty::add(const basyx::object & newValue)
{
  this->getProxy()->createElement(IProperty::Path::Value, newValue);
}

void ConnectedCollectionProperty::remove(basyx::object & objectRef)
{
  this->getProxy()->deleteElement(IProperty::Path::Value, objectRef);
}

basyx::object::object_list_t ConnectedCollectionProperty::getElements() const
{
  return this->retrieveObject().Get<basyx::object::object_list_t>();
}

int ConnectedCollectionProperty::getElementCount() const
{
  return this->retrieveObject().Get<basyx::object::object_list_t>().size();
}

}
}
